cmd/engflow_auth/main.go

```diff
@@ -73,12 +75,39 @@ func (r *appState) build(cliCtx *cli.Context) error {
 		r.browserOpener = &browser.StderrPrint{}
 	}
 	if r.tokenStore == nil {
-		tokenStore, err := oauthtoken.NewKeyring()
+		keyring, err := oauthtoken.NewKeyring()
 		if err != nil {
-			return autherr.CodedErrorf(autherr.CodeTokenStoreFailure, "failed to open token store: %w", err)
+			return autherr.CodedErrorf(autherr.CodeTokenStoreFailure, "failed to open keyring-based token store: %w", err)
 		}
 
-		r.tokenStore = oauthtoken.NewCacheAlert(tokenStore, cliCtx.App.ErrWriter)
+		userConfigDir, err := os.UserConfigDir()
+		if err != nil {
+			return autherr.CodedErrorf(autherr.CodeTokenStoreFailure, "failed to discover user's config dir: %w", err)
+		}
+		tokensDir := filepath.Join(userConfigDir, "engflow_auth", "tokens")
+		fileStore, err := oauthtoken.NewFileTokenStore(tokensDir)
+		if err != nil {
+			return autherr.CodedErrorf(autherr.CodeTokenStoreFailure, "failed to open file-based token store: %w", err)
+		}
+
+		var writeStore oauthtoken.LoadStorer
+		switch writeStoreName := cliCtx.String("store"); writeStoreName {
```

Comment:
Does this work for subcommands that don't have a `-store` flag? I'd expect `cliCtx.String("store")` to return `""` for those.

Reply:
Let's move the `-store` flag to the subcommands. It only makes sense in the context of a couple of the subcommands. That does make constructing this harder, but form should follow function.

Reply:
Are you saying that `appState.tokenStore` should be conditionally constructed per subcommand? This means:
* each subcommand has a different `appState`, which does make the app more complicated
* let's suppose a caller wants to make multiple invocations while changing the `-store` flag - they need to know which subcommands require/don't require the flag, rather than unconditionally prefixing all commands - which makes their logic more complicated as well

I'll put up the change so that we can see the difference, but my initial guess is that moving the store flag to the root is both more easily implemented and conceptualized by the caller

Reply:
I'm not stipulating too much about how it's implemented, I only disagree that `-store` should be a global flag. From a user interface perspective, it only makes sense on the `login` and `import` commands. For `help` and `version`, it's meaningless. For `get`, `export`, and `logout`, it could be misleading.

This architecture makes it quite difficult to take the subcommand's context into account. I'm not sure how you can do that without either modifying `appState` or constructing it per subcommand. `Fallback` is coupled with the application logic.

Reply:
Actually how about this: in `appState.build`, we have a `*cli.Context`. That method is called as the `cli.App.Before` hook, and the documentation for `App` says that's called after the context is ready but before the subcommand implementation. I think that means the subcommand flags are parsed already.

So I think the only additional thing we need here is to check whether the flag is defined for the current subcommand. If it's not defined (e.g., for `version`), then we can default to keyring. If it is defined, then returning `""` means the user explicitly passed in `""`, and it should be an error.

Reply:
Hey, that works well! Glad we don't need to shard construction across `Before` actions of subcommands as well :)

> If it is defined, then returning "" means the user explicitly passed in "", and it should be an error

The other option is a subcommand is added/modified to write to token storage, but the flag is not added to that subcommand. In this case, we could make `""` default to a storage backend that always errors, so that this issue can get caught by us early - though this does mean if the user passes `--store=''` they will get a confusing error message back. Which do you think is better? (I think passing `--store=''` is an insane thing to do, and think the check that there's a `--store` flag on every subcommand that writes tokens provides more value - but don't feel strongly this way)

Reply:
...actually, we may be able to have it both ways. Let me see what I can do here

Reply:
Let's land this is, I'm eager to get it set up in dogfood.

Mild preference: if a `-store` flag is defined for the subcommand, it should be an immediate error to set it to `""`. It's too bad this CLI library doesn't distinguish between "no flag" and "flag set to the empty string", but we at least have `Context.FlagNames`. Agree though that it's unlikely to happen accidentally, so I don't care much if we fix that.

Reply:
Fixed! I attached a validator function to the flag, so when the flag is present on a subcommand, the validator ensures it has a legal value. If the flag value is empty during construction, it is therefore only due to the subcommand not having the `--store` flag present - so we have an erroring store backend (for writes only) which should enforce that future subcommands that want to write to the token store have `--store` added.

Only question - since `--store` only affects where token writes happen - do you think we should change the name to reflect that it controls the token write path only?